### ISSUE TITLE
POST test fails in Rspec/Capybara but works in browser, similar test works fine.

### ISSUE THREAD

**Issue by Randall-Coding** [teamcapybara/capybara#2145]
**Note:** This is for issues with Capybara.  If you have a howto type question, please ask on the mailing list as requested in the README: http://groups.google.com/group/ruby-capybara

## Meta
Capybara Version:
3.12.0

Driver Information (and browser if relevant):
Firefox 64bit for Linux 1.0
Selenium-webdriver 3.141.0

## Expected Behavior
The page is expected to be populated with data of a given transaction.  And in my normal web browser it does work.  Only this one test is failing.

When I run my server (after clearing the database) it looks like this in the browser  
![ropsten_success](https://user-images.githubusercontent.com/39175191/50484118-2453b000-09ac-11e9-9e9f-e61f4229ce5c.png)

## Actual Behavior

When I test this feature, I receive this
> Failures:
> 
>   1) Search page performs search with Ropsten displays the correct data
>      Got 5 failures:
> 
>      1.1) Failure/Error: expect(page.has_content?("0x6f61ee4dc528443d8b6ab12240a53588fb72c659e1abfe6e6cc0d2346359be4a")).to be true
>           
>             expected true
>                  got false
>           # ./spec/features/basic_searches_spec.rb:21:in `block (3 levels) in <top (required)>'
> 
>      1.2) Failure/Error: expect(page.has_content?("Success")).to be true
>           
>             expected true
>                  got false
>           # ./spec/features/basic_searches_spec.rb:22:in `block (3 levels) in <top (required)>'
> 
>      1.3) Failure/Error: expect(page.has_content?("4588323")).to be true
>           
>             expected true
>                  got false
>           # ./spec/features/basic_searches_spec.rb:23:in `block (3 levels) in <top (required)>'
> 
>      1.4) Failure/Error: expect(page.has_content?("0xbdfeff9a1f4a1bdf483d680046344316019c58cf")).to be true
>           
>             expected true
>                  got false
>           # ./spec/features/basic_searches_spec.rb:24:in `block (3 levels) in <top (required)>'
> 
>      1.5) Failure/Error: expect(page.has_content?("21000")).to be true
>           
>             expected true
>                  got false
>           # ./spec/features/basic_searches_spec.rb:25:in `block (3 levels) in <top (required)>'
> 
> Finished in 2.78 seconds (files took 1.03 seconds to load)
> 6 examples, 1 failure

## Steps to reproduce

The Rspec/Capybara code in question is as follows

    context "performs search with Ropsten" do
      before(:all) do
        visit root_path
      end
    it "displays the correct data" do
      fill_in "search-transaction" , with: "0x6f61ee4dc528443d8b6ab12240a53588fb72c659e1abfe6e6cc0d2346359be4a"
      choose "Ropsten"
      click_button "Search"
      expect(current_path).to eq(root_path)
      expect(page.has_content?("0x6f61ee4dc528443d8b6ab12240a53588fb72c659e1abfe6e6cc0d2346359be4a")).to be true
      expect(page.has_content?("Success")).to be true
      expect(page.has_content?("4588323")).to be true
      expect(page.has_content?("0xbdfeff9a1f4a1bdf483d680046344316019c58cf")).to be true
      expect(page.has_content?("21000")).to be true
      end
    end

Here is the tricky part though.  I have confirmed the radio button is being selected and the params passed are correct as shown here

> My params: 
> <ActionController::Parameters {"hash"=>"0x6f61ee4dc528443d8b6ab12240a53588fb72c659e1abfe6e6cc0d2346359be4a", "mode"=>"ropsten", "authenticity_token"=>"bFdyLUHmN4bXzdUGlOh2QsIHe9kojYHvhr1Rp/UGD+2OuGId1Kf7KyHh7Vf3g967FsOuc5aetw5odePASR+uuw==", "controller"=>"transaction", "action"=>"index"} permitted: false>
> 

The important parameters are "mode" and "hash".  

To make matters worse the next test that looks nearly identical passes.  The passing test is shown below
The only difference that I can see is that it is "Mainnet" instead of "Ropsten"

    context "performs search with Mainnet" do
      before(:all) do
        visit root_path
    end
    it "displays the correct data" do
      fill_in "search-transaction" , with: "0x4f85d4e93eaa5cd77fd79671b175b380cfc1f8d9635990c14bddb8725efa5250"
      choose "Mainnet"
      click_button "Search"
      expect(current_path).to eq(root_path)
      expect(page.has_content?("0x4f85d4e93eaa5cd77fd79671b175b380cfc1f8d9635990c14bddb8725efa5250")).to be true
      expect(page.has_content?("0x6b70b1ac178b2ce967673223d7321cb5de9da657")).to be true
      expect(page.has_content?("Success")).to be true
      expect(page.has_content?("6849773")).to be true
      expect(page.has_content?("0xf00d27c64e125aa2da41298edd3660ac077d1e91")).to be true
      expect(page.has_content?("21000")).to be true
     end
    end

I wonder if this has to do with the API call I am making to Etherscan?  Is Rspec or Capybara timing out perhaps?  

### Github source code
https://github.com/Randall-Coding/ethereum-api-2.0

### HTML 
Relevant HTML 
```
     <!-- Search bar and button -->
      <div class="searchbar">
        <form action="/" method="post">
           <div class="search-tbox">
        Transaction # <input id="search-transaction" name="hash" type="text" placeholder="ex. 0x6f61ee4dc528443d8b6ab12240a53588fb72c659e1abfe6e6cc0d2346359be4a"
          value="  0x6f61ee4dc528443d8b6ab12240a53588fb72c659e1abfe6e6cc0d2346359be4a    " >  </input>
      </div>
      <div class="mode">
        <a href="javascript:void(0)" class='examples'> Example numbers </a>
        <fieldset class="mode">
            <input type= "radio" value="mainnet" name="mode" id="mode-mainnet" >
              <label for="mode-mainnet"> Mainnet </label>
           </input>
            <input type="radio" value="ropsten" name="mode"  id="mode-ropsten" checked>
              <label for="mode-ropsten"> Ropsten  </label>
            </input>
        </fieldset>
        <button id="btn-search1" class="search" type="submit"> Search </button>
        <input type="hidden" name="authenticity_token" id="authenticity_token" value="JKZpoYBdDSFYl4uqKBP/Y3yNun06RBWpAxbVBXsBr5AtwdfgMY6H47FpdCTi5giXZMIPaWryekOWF+OpkvjiuQ==" />
         </div>
       </form>
     </div>
```

**Comment by twalpole** [comment]
Assuming you are actually using the Selenium driver for these tests it's definitely possible  the Ropsten network is responding slower than Mainnet and that you don't have `Capybara.default_max_wait_time` set high enough for the tests you're writing.  Unfortunately due to your use of the `eq` matcher, rather than the matchers provided by Capybara, the error messages you are getting provide no useful information. If instead you used the matchers provided by Capybara your error messages would actually be useful and your tests overall would be more stable and read better


     it "displays the correct data" do
        fill_in "search-transaction" , with: "0x4f85d4e93eaa5cd77fd79671b175b380cfc1f8d9635990c14bddb8725efa5250"
        choose "Mainnet"
        click_button "Search"
        expect(page).to have_current_path(root_path)
        expect(page). to have_content("0x4f85d4e93eaa5cd77fd79671b175b380cfc1f8d9635990c14bddb8725efa5250")
        expect(page).to have_content("0x6b70b1ac178b2ce967673223d7321cb5de9da657")
        expect(page).to have_content("Success")
        expect(page).to have_content("6849773")
        expect(page).to have_content("0xf00d27c64e125aa2da41298edd3660ac077d1e91")
        expect(page).to have_content("21000")
     end

**Comment by Randall-Coding** [comment]
Oh thanks for the quick answer.  I managed to convince myself no one had responded (I scrolled down too fast and must have mistaken your comment for my own post).  Should have checked my Github associated e-mail, doh!

But anyway I just tried your solution but adding `Capybara.default_max_wait_time = 20` to the spec.  Still the same output.  I also tried a max wait of 40 still the same output.

**Comment by twalpole** [comment]
Did you change the expects so you get useful error messages?  As for the wait time, that would only change anything if you were actually using Selenium - I’m guessing you’re actually using the rack_test driver (full stacktrace from one of the failures would confirm that)

**Comment by Randall-Coding** [comment]
Was just testing those Capybara matchers now.  One moment

**Comment by Randall-Coding** [comment]
I used `bundle exec rspec spec/features/basic_searches_spec.rb  -b` to provide a backtrace.  

Failures:

  1) Search page performs search with Ropsten displays the correct data
     Got 6 failures:

     1.1) Failure/Error: expect(find('#overview')).to have_content('Success')
            expected to find text "Success" in "Ethereum transaction information\n \nCould not obtain data. Try a different transaction hash"
          # /var/lib/gems/2.5.0/gems/rspec-expectations-3.8.2/lib/rspec/expectations/failure_aggregator.rb:49:in `call'
          # /var/lib/gems/2.5.0/gems/rspec-support-3.8.0/lib/rspec/support.rb:106:in `notify_failure'
          # /var/lib/gems/2.5.0/gems/rspec-expectations-3.8.2/lib/rspec/expectations/fail_with.rb:35:in `fail_with'
          # /var/lib/gems/2.5.0/gems/rspec-expectations-3.8.2/lib/rspec/expectations/handler.rb:40:in `handle_failure'
          # /var/lib/gems/2.5.0/gems/rspec-expectations-3.8.2/lib/rspec/expectations/handler.rb:50:in `block in handle_matcher'
          # /var/lib/gems/2.5.0/gems/rspec-expectations-3.8.2/lib/rspec/expectations/handler.rb:27:in `with_matcher'
          # /var/lib/gems/2.5.0/gems/rspec-expectations-3.8.2/lib/rspec/expectations/handler.rb:48:in `handle_matcher'
          # /var/lib/gems/2.5.0/gems/rspec-expectations-3.8.2/lib/rspec/expectations/expectation_target.rb:65:in `to'
          # ./spec/features/basic_searches_spec.rb:20:in `block (3 levels) in <top (required)>'
          # /var/lib/gems/2.5.0/gems/rspec-core-3.8.0/lib/rspec/core/example.rb:254:in `instance_exec'
          # /var/lib/gems/2.5.0/gems/rspec-core-3.8.0/lib/rspec/core/example.rb:254:in `block in run'
          # /var/lib/gems/2.5.0/gems/rspec-core-3.8.0/lib/rspec/core/example.rb:500:in `block in with_around_and_singleton_context_hooks'
          # /var/lib/gems/2.5.0/gems/rspec-core-3.8.0/lib/rspec/core/example.rb:457:in `block in with_around_example_hooks'
          # /var/lib/gems/2.5.0/gems/rspec-core-3.8.0/lib/rspec/core/hooks.rb:464:in `block in run'
          # /var/lib/gems/2.5.0/gems/rspec-core-3.8.0/lib/rspec/core/hooks.rb:604:in `block in run_around_example_hooks_for'
          # /var/lib/gems/2.5.0/gems/rspec-core-3.8.0/lib/rspec/core/example.rb:342:in `call'
          # /var/lib/gems/2.5.0/gems/rspec-rails-3.8.1/lib/rspec/rails/adapters.rb:127:in `block (2 levels) in <module:MinitestLifecycleAdapter>'
          # /var/lib/gems/2.5.0/gems/rspec-core-3.8.0/lib/rspec/core/example.rb:447:in `instance_exec'
          # /var/lib/gems/2.5.0/gems/rspec-core-3.8.0/lib/rspec/core/example.rb:447:in `instance_exec'
          # /var/lib/gems/2.5.0/gems/rspec-core-3.8.0/lib/rspec/core/hooks.rb:373:in `execute_with'
          # /var/lib/gems/2.5.0/gems/rspec-core-3.8.0/lib/rspec/core/hooks.rb:606:in `block (2 levels) in run_around_example_hooks_for'
          # /var/lib/gems/2.5.0/gems/rspec-core-3.8.0/lib/rspec/core/example.rb:342:in `call'
          # /var/lib/gems/2.5.0/gems/rspec-expectations-3.8.2/lib/rspec/expectations/failure_aggregator.rb:10:in `block in aggregate'
          # /var/lib/gems/2.5.0/gems/rspec-support-3.8.0/lib/rspec/support.rb:113:in `with_failure_notifier'
          # /var/lib/gems/2.5.0/gems/rspec-expectations-3.8.2/lib/rspec/expectations/failure_aggregator.rb:8:in `aggregate'

**Comment by Randall-Coding** [comment]
My apologies for the formatting.  Not sure how to escape all that.  

As for the new matchers, it provides information I was already aware of by printing the `page.body` to stdout.  What this error message means is that my API call turned up `nil`

**Comment by Randall-Coding** [comment]
I will check tomorrow if I am using Selenium (I know I have the gem installed but that is about it).  Again sorry for the delay, and thanks for the advice.

**Comment by twalpole** [comment]
Unfortunately the backtrace doesn't show which driver you're using, possibly because you're using aggregate failures (which really don't make a lot of sense in feature tests where later expectations  generally depend on prior expectations having succeeded).  Additionally, `expect(find('#overview')).to have_content('Success')` should normally be written as `expect(page).to have_css('#overview', text: 'Success')` for performance reasons.

As for the error message, due to the text "Could not obtain data. Try a different transaction hash"  it seems pretty clear that the hash you're searching for isn't in whatever data the search is being submitted to.  This doesn't really seem to be a Capybara issue.  I'm going to close this for now, but if you can provide info showing that it is actually a Capybara issue I'll reopen.

**Comment by Randall-Coding** [comment]
Ok I confirmed that I was using rack_test driver (using `Capybara.current_driver`) and not Selenium, so I changed my default driver to Selenium.  Then I got an error that I needed geckodriver which should be placed in PATH.  I got that working. 
Capybara.default_max_wait_time = 20

Now I get the same failure, it just takes longer to fail :)  I also tried :selenium_chrome with the same results.

**Comment by Randall-Coding** [comment]
> "Try a different transaction hash" it seems pretty clear that the hash you're searching for isn't in whatever data the search is being submitted to.

The image I provided above shows that the data is in fact in the data base, and it works on my normal browser.  Here is the image again 
![Uploading ropsten_success.png…]()

This can also be confirmed using the version I deployed on my website https://www.randallcoding.com/api/ethereum  make sure to select 'ropsten'.

As I demonstrated in my debugging the params being displayed by my controller show "mode" => 'ropsten' as desired.  If "mode" => "mainsten" we would expect the error message (or if the hash was not valid for either), however this does not appear to be the case.

**Comment by Randall-Coding** [comment]
I also ran the two tests in isolation using 

`bundle exec rspec spec/features/basic_searches_spec.rb -t ropsten`  for the failing test. 
and
`bundle exec rspec spec/features/basic_searches_spec.rb -t mainnet' for the passing test.

Same results, so it doesn't seem to be order dependent.  This is really perplexing because both are doing almost exactly the same thing except one selects a different radio button and calls a different function on the API.  And they both work in real life, but the first one does not work in Capybara :( 

Wish I could nail down this issue better for you guys, but I am at a loss here.

**Comment by Randall-Coding** [comment]
Ok I solved it.  Clearing the test database before hand fixed the issue.  I should have thought of this earlier, but I thought it must be a Capybara config problem.  I ruled out a database issue since my test runs showed no problem with using cached data. 

So looks like it was something to do with the test database and perhaps some left over fixtures in there?  
Looks like not a Capybara issue after all (I think?).  Ah the joys of testing.
